content/posts/2025-09-01-gsoc-ubsan-trap-messages.md

```diff
@@ -0,0 +1,410 @@
+(lldb) r
+Process 17347 launched: 'add' (arm64)
+Process 17347 stopped
+* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BREAKPOINT (code=1, subcode=0x100003d3c)
+    frame #0: 0x0000000100003d3c add`add(a=2147483647, b=1) at add.c:5:13
+   2    #include <stdio.h>
+   3   
+   4    int add(int a, int b) {
+-> 5        return a + b;
+   6    }
+   7   
+   8    int main(int argc, const char** argv) {
+(lldb) bt
+* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BREAKPOINT (code=1, subcode=0x100003d3c)
+  * frame #0: 0x0000000100003d3c add`add(a=2147483647, b=1) at add.c:5:13
+    frame #1: 0x0000000100003eec add`main(argc=3, argv=0x000000016fdff110) at add.c:13:18
+    frame #2: 0x00000001842bab98 dyld`start + 6076
+(lldb) dis -p
+add`add:
+->  0x100003d3c <+40>: brk    #0x5500
+    0x100003d40 <+44>: ldr    w0, [sp, #0x4]
+    0x100003d44 <+48>: add    sp, sp, #0x10
+    0x100003d48 <+52>: ret   
+```
+
+
+We can see that a `brk` instruction was hit while handling the `a + b` expression but there is no good explanation of what happened. `brk` is the trap instruction on arm64 but it is not particularly clear this has anything to do with UBSan. For this toy example we can speculate that integer overflow occurred because the program was built with trapping UBSan and the trap was hit while handling `a + b`. However, in real programs built with trapping UBSan and potentially other hardening mechanisms it is often far less obvious what happened.
+
+
+For this particular example. The information that this is an integer overflow UBSan check is actually there but it is not very obvious. On x86_64 and arm64 the reason for trapping is actually [encoded in the operand to the trap instruction](https://maskray.me/blog/2023-01-29-all-about-undefined-behavior-sanitizer ). In this case the `#0x5500` immediate to the brk instruction encodes that this is a UBSan trap for integer overflow. The UBSan immediate is encoded as `('U' << 8) + SanitizerHandler` where `SanitizerHandler` is the enum value from the `SanitizerHandler` enum inside Clang’s internals.
+
+
+As we can see the debugging experience with UBSan traps is not ideal and improving this was the primary goal of the GSoC project.
+
+
+
+
+## Human readable descriptions of UBSan traps in LLDB
+
+
+During my GSoC project I implemented support for displaying human readable descriptions of UBSan traps in LLDB to improve the debugging experience.
+
+
+The approach used is based on how `__builtin_verbose_trap` is currently implemented inside Clang [11] [12]. `__builtin_verbose_trap` was implemented in the past for [libc++ hardening](https://discourse.llvm.org/t/rfc-hardening-in-libc/73925). At a high-level this works by encoding the reason for trapping as a string on the trap instruction in the debug info for the program being compiled. Then when a trap is hit in the debugger, the debugger retrieves this string and shows it as the reason for trapping.
```

Comment:
```suggestion
The approach used is based on how `__builtin_verbose_trap` is currently implemented inside Clang [11] [12]. `__builtin_verbose_trap` was implemented in the past for [libc++ hardening](https://discourse.llvm.org/t/rfc-hardening-in-libc/73925). At a high-level this works by encoding the reason for trapping on the trap's debug info. When a trap is hit in the debugger, the debugger retrieves this string and shows it as the reason for trapping.
```

Reply:
Should we mention that you are going to expand on this in an upcoming section? Currently it sounds like you are glossing over the implementation details, although you *do* explain them later on

Reply:
Actually, can we just remove this entire paragraph (or move the important bits into the `Encoding the trap reason in the debug info` subsection below?

I feel like gives us following clear structure:
1. "We want to make the trap reasons human readable in the debugger"
2. "We could make the debugger handle this without compiler support"
  * "These are the reasons it didn't work"
3. "Encode reason in debug info"
  * "This is the approach we took"